docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
 requests that affected them and the ABCI responses they affected. The `ABCIListener` interface also exposes a
-`ListenSuccess` method which is (optionally) used by the `BaseApp` to await positive acknowledgement of message
-receipt from the `StreamingService`.
+`ListenKVStorePair` method which is used by the `StoreKVPairWriteListener` to stream state change events.
 
 ```go
 // ABCIListener interface used to hook into the ABCI message processing of the BaseApp
 type ABCIListener interface {
-    // ListenBeginBlock updates the streaming service with the latest BeginBlock messages
-    ListenBeginBlock(ctx types.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
-    // ListenEndBlock updates the steaming service with the latest EndBlock messages
-    ListenEndBlock(ctx types.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
-    // ListenDeliverTx updates the steaming service with the latest DeliverTx messages
-    ListenDeliverTx(ctx types.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
-    // HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
-    // in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. When `false, the app will operate in fire-and-forget mode.
-    // When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
-    // be replayed to all listeners when the node restarts and all successful listeners that received data
-    // prior to the halt will receive duplicate data. Whether or not a listener operates in a fire-and-forget mode
-    // is determined by the listener's configuration property `halt_app_on_delivery_error = true|false`.
-    HaltAppOnDeliveryError() bool
-}
-
-// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
```

Comment:
These methods should take a `context.Context` as a first parameter, so that they can be canceled by callers.

Reply:
I'm not well versed in the best use of the `context.Context` and my understanding is that when it comes to  making client calls, you are supposed to create a new `context` for each call that you make. This is how the implementation looks like. It's based on the `go-plugin` examples. 

https://github.com/cosmos/cosmos-sdk/blob/c0871cd41487ed76b0947553d3195ff6045a478f/streaming/plugins/abci/grpc_abci_v1/grpc.go#L19-L26

Reply:
Not exactly, no, and the code you linked is also incorrect. (Usually, if you see context.Background, it's a red flag that something is wrong.)

The main purpose of context.Context is to allow callers to interrupt the code they call. For that to work, you have to accept a context from your callers, so that your code will respect their cancellation.

Contexts are usually created at the start of a request's lifetime, which almost always happens in the transport's client and server types rather than the application. (Transport here being gRPC or HTTP.) They control the request lifecycle, your application just handles it.